cmd/thanos/receive_route.go

```diff
@@ -0,0 +1,222 @@
+			*tenantHeader,
+			*defaultTenantID,
+			*replicationFactor,
+			time.Duration(*forwardTimeout),
+			component.Receive,
+		)
+	})
+}
+
+func runReceiveRoute(
+	g *run.Group,
+	logger log.Logger,
+	reg *prometheus.Registry,
+	tracer opentracing.Tracer,
+	httpBindAddr string,
+	httpGracePeriod time.Duration,
+	rwAddress string,
+	rwServerCert string,
+	rwServerKey string,
+	rwServerClientCA string,
+	rwClientCert string,
+	rwClientKey string,
+	rwClientServerCA string,
+	rwClientServerName string,
+	cw *route.ConfigWatcher,
+	tenantHeader string,
+	defaultTenantID string,
+	replicationFactor uint64,
+	forwardTimeout time.Duration,
+	comp component.SourceStoreAPI,
+) error {
+	logger = log.With(logger, "component", "receive-route")
+	level.Warn(logger).Log("msg", "setting up receive-route")
+	rwTLSConfig, err := tls.NewServerConfig(log.With(logger, "protocol", "HTTP"), rwServerCert, rwServerKey, rwServerClientCA)
+	if err != nil {
+		return err
+	}
+	dialOpts, err := extgrpc.StoreClientGRPCOpts(logger, reg, tracer, rwServerCert != "", rwClientCert, rwClientKey, rwClientServerCA, rwClientServerName)
+	if err != nil {
+		return err
+	}
+
+	webHandler := route.NewHandler(log.With(logger, "component", "receive-handler"), &route.Options{
+		ListenAddress:     rwAddress,
+		Registry:          reg,
+		TenantHeader:      tenantHeader,
+		DefaultTenantID:   defaultTenantID,
+		ReplicationFactor: replicationFactor,
+		Tracer:            tracer,
+		TLSConfig:         rwTLSConfig,
+		DialOpts:          dialOpts,
+		ForwardTimeout:    forwardTimeout,
+	})
+
+	httpProbe := prober.NewHTTP()
+	statusProber := prober.Combine(
+		httpProbe,
+		prober.NewInstrumentation(comp, logger, extprom.WrapRegistererWithPrefix("thanos_", reg)),
+	)
+
+	// Start all components while we wait for TSDB to open but only load
+	// initial config and mark ourselves as ready after it completed.
+	level.Debug(logger).Log("msg", "setting up hashring")
+	{
+		// Note: the hashring configuration watcher
+		// is the sender and thus closes the chan.
+		// In the single-node case, which has no configuration
+		// watcher, we close the chan ourselves.
+		updates := make(chan route.Hashring, 1)
+
+		// Check the hashring configuration on before running the watcher.
+		if err := cw.ValidateConfig(); err != nil {
+			cw.Stop()
+			close(updates)
+			return errors.Wrap(err, "failed to validate hashring configuration file")
+		}
+		{
+			ctx, cancel := context.WithCancel(context.Background())
+			g.Add(func() error {
+				return route.HashringFromConfigWatcher(ctx, updates, cw)
+			}, func(error) {
+				cancel()
+			})
+		}
+		cancel := make(chan struct{})
+		g.Add(func() error {
+			for {
+				select {
+				case h, ok := <-updates:
+					if !ok {
+						return nil
+					}
+					msg := "hashring has changed; server is not ready to receive web requests"
+					statusProber.NotReady(errors.New(msg))
+					level.Info(logger).Log("msg", msg)
+					webHandler.Hashring(h)
+					statusProber.Ready()
```

Comment:
Added the ready prober here,
Need to check if the change makes sense 🤔 cc @squat

Reply:
I'm sure this is correct.
It seems like I miss the part that connects hasring updates to TSDB flush/open.